tools/c7n_mailer/c7n_mailer/ldap_lookup.py

```diff
@@ -0,0 +1,146 @@
+        email_to_addr = []
+        ldap_user_metadata = ldap_user_to_email_map.get(ldap_username)
+        user_email = ldap_user_metadata.get(self.email_key, None)
+        if user_email:
+            email_to_addr.append(user_email)
+        if email_manager:
+            user_manager_dn = ldap_user_metadata.get('manager', None)
+            if user_manager_dn:
+                ldap_user_manager_metadata = ldap_user_to_email_map.get(user_manager_dn, {})
+                user_manager_email = ldap_user_manager_metadata.get(self.email_key, None)
+                if user_manager_email:
+                    email_to_addr.append(user_manager_email)
+        return email_to_addr
+
+    def get_dict_from_ldap_object(self, ldap_user_object):
+        ldap_user_metadata = {attr.key: attr.value for attr in ldap_user_object}
+        ldap_user_metadata['dn'] = ldap_user_object.entry_dn
+        ldap_user_metadata[self.email_key] = ldap_user_metadata[self.email_key].lower()
+        return ldap_user_metadata
+
+    def get_metadata_from_uid(self, uid):
+        if self.uid_regex:
+            # for example if you set ldap_uid_regex in your mailer.yml to "^[0-9]{6}$" then it
+            # would only query LDAP if your string length is 6 characters long and only digits.
+            # re.search("^[0-9]{6}$", "123456")
+            # Out[41]: <_sre.SRE_Match at 0x1109ab440>
+            # re.search("^[0-9]{6}$", "1234567") returns None, or "12345a' also returns None
+            if not re.search(self.uid_regex, uid):
+                regex_msg = 'uid does not match regex: %s %s' % (self.uid_regex, uid)
+                self.log.debug(regex_msg)
+                return None
+        if self.cache_engine:
+            cache_result = self.caching.get(uid, None)
+            if cache_result:
+                cache_msg = 'Got ldap metadata from local cache for: ' % uid
+                self.log.debug(cache_msg)
+                return cache_result
+        ldap_filter = '(%s=%s)' % (self.uid_key, uid)
+        self.connection.search(self.base_dn, ldap_filter, attributes=self.attributes)
+        if len(self.connection.entries) == 0:
+            self.log.warning("userid not found %s", uid)
+            return None
+        if len(self.connection.entries) > 1:
+            self.log.warning("too many results for search %s", ldap_filter)
+            return None
+        ldap_user_metadata = self.get_dict_from_ldap_object(self.connection.entries[0])
+        if self.cache_engine:
+            self.caching.set(uid, ldap_user_metadata)
+        return ldap_user_metadata
+
+
+# the purpose of this class is to create a light wrapper around a standard library called shelve
+# which can store key/value pairs to a local file (using pickle), similar to how memcache or redis
+# stores things. This normalizes the methods to set/get a function, so you can interchangeable
+# decide which caching system to use, a local file, or memcache, redis, etc
+class LocalShelf(object):
```

Comment:
If someone doesn't want to deal with redis as a dependency, they can just use a local file. This shouldn't be used in lambda, but works great say if you're calling the mailer directly from your container or EC2 instance. all the methods in this object are normalized so they're compatible with redis or memcache.